### DISCUSSION TITLE
Feedback: Paste Layout components should be more generic

### DISCUSSION THREAD

**Discussion by akmittal** [twilio-labs/paste#2637] [General] [answered]
I feel Paste's Layout components like Flex and Stack are missing quite a few features.

1. `Box` component support `rowGap` and `columnGap` which `Flex` does not. `rowGap` and `columnGap` are essentially flexbox properties. So ideally `Flex` should support these.
2. `Stack` component has a spacing property for the fixed gap, But when spacing needs to be dynamic (like `space-between`, `space-around` ). We need to resort to `Flex`.
3. `Stack` component should support properties like `margin` and `padding`. That will remove the need to wrap component in another container component. Which leads to unnecessary DOM nesting

**Comment by SiTaggart** [comment, marked as answer]
Hi @akmittal 

Thanks for the feedback! I'll tackle each item one by one.

1. Yeah, `Flex` is very high on our least liked component, internally and externally. It was an easrly component and we've learnt a lot since then. We would love to revisit it from the ground up when we get a chance to do so. With that said I don't think we'd want to add anything new to it in its current state, because we probably want to fundamentally change how it works.
2. That's an interesting observation. `Stack` was always designed to be the simplest, single purpose  component out there. Just place a uniform amount of space between each child. So I don't necessarily mind that you reach for a different component when your use case is more than what it supports, it's just unfortunate that `Flex` is the thing that you have to reach for. 
3. That's also an interesting observation. We've always leaned towards leaving layout concerns to wrappers so that each component doesn't take up more space than it is responsible for. But Stack being a layout component might make this an pretty reasonable suggestion that I'll present to the team.

Something to consider too, we recently extended `Box` to use the CSS grid spec, which can be used to create even more interesting layouts in far less code. It might be an option you end up prefering.

We're also considering adding a CSS grid based version of the `Grid` component too, to make layouts even more flexible.

Thanks again for the feedback, and let us know what you think about `Box` and CSS grid.
